src/track/globaltrackcache.cpp

```diff
@@ -30,240 +36,356 @@ TrackRef createTrackRef(const Track& track) {
+        const TrackRef trackRef = createTrackRef(*plainPtr);
+        VERIFY_OR_DEBUG_ASSERT(!trackRef.getCanonicalLocation().isEmpty()) {
             return false;
         }
-        VERIFY_OR_DEBUG_ASSERT(createTrackRef(*(*i).plainPtr) == trackRef) {
+        VERIFY_OR_DEBUG_ASSERT(trackRef.getCanonicalLocation() == (*i).first) {
             return false;
         }
-        VERIFY_OR_DEBUG_ASSERT(1 == m_tracksByCanonicalLocation.count(canonicalLocation)) {
+        VERIFY_OR_DEBUG_ASSERT(1 == m_tracksByCanonicalLocation.count(trackRef.getCanonicalLocation())) {
             return false;
         }
-        TracksById::const_iterator j(
-                m_tracksById.find(trackId));
+        TracksById::const_iterator j = m_tracksById.find(trackRef.getId());
         VERIFY_OR_DEBUG_ASSERT(
-                (m_tracksById.end() == j) || ((*j).ref == trackRef)) {
+                (m_tracksById.end() == j) || ((*j).second == plainPtr)) {
             return false;
         }
     }
     return true;
 }
 
-GlobalTrackCacheLocker GlobalTrackCache::lookupById(
-        const TrackId& trackId) const {
-    GlobalTrackCacheLocker cacheLocker;
-    if (trackId.isValid()) {
-        const TrackPointer pTrack(lookupInternal(trackId));
-        if (pTrack) {
-            cacheLocker.m_lookupResult = GlobalTrackCacheLookupResult::HIT;
-            cacheLocker.m_trackRef = createTrackRef(*pTrack);
-            cacheLocker.m_pTrack = pTrack;
-        } else {
-            cacheLocker.m_lookupResult = GlobalTrackCacheLookupResult::MISS;
+void GlobalTrackCache::deactivate() {
+    DEBUG_ASSERT(verifyConsistency());
+    // Ideally the cache should be empty when destroyed.
+    // But since this is difficult to achieve all remaining
+    // cached tracks will evicted no matter if they are still
+    // referenced or not. This ensures that the eviction
+    // callback is triggered for all modified tracks before
+    // exiting the application.
+    while (!m_indexedTracks.empty()) {
+        evictAndDelete(
+                m_indexedTracks.begin(),
+                true);
+    }
+    // Verify that all indexed tracks have been evicted
+    DEBUG_ASSERT(m_tracksById.empty());
+    DEBUG_ASSERT(m_tracksByCanonicalLocation.empty());
+    // The singular cache instance is already unavailable and
+    // all unindexed tracks will simply be deleted when their
+    // shared pointer goes out of scope. Their modifications
+    // will be lost.
+    m_unindexedTracks.clear();
+    m_pDeleter = nullptr;
+}
+
+void GlobalTrackCache::reset() {
+    if (!m_indexedTracks.empty()) {
+        kLogger.warning()
+                << "Resetting indices while"
+                << m_indexedTracks.size()
+                << "tracks are still cached and indexed";
+        for (auto i = m_indexedTracks.begin(); i != m_indexedTracks.end(); ++i) {
+            Track* plainPtr = (*i).first;
+            DEBUG_ASSERT(m_unindexedTracks.find(plainPtr) == m_unindexedTracks.end());
```

Comment:
Is here an insert missing? This will finally delete all track objects without saving. Is this intended?

Reply:
Yes, the insert must have been lost on the way. Strange. Do you plan to re-add this while already preparing a fix?

Reply:
I think just reseting th cache is wrong anyway. This happens when a user change one of the library root directories. So only some tracks are effected by this. 
This has to be done more intelligent to not loose user edits.

Reply:
I think it is better to swap the responsibility of "m_indexedTracks". It should be used only for keeping track of the pointer livetime to back up the the time between releasing the last reference and deleting. 

The other two container should two containers should use weak pointers.